
<!DOCTYPE html>
<html>
  <head>
    <title>MG-RAST API</title>
    <meta http-equiv="content-type" content="text/html; charset=utf-8" />
    <script type="text/javascript" src="bootstrap.min.js"></script>
    <link rel="stylesheet" type="text/css" href="bootstrap.min.css" />
  </head>

  <body style="background-color: white; margin: 15px;">
    <div class="navbar navbar-inverse navbar-fixed-top">
      <div class="navbar-inner">
	<div class="container">
	  <a class="brand" href="./index.html">MG-RAST API documentation</a>
	</div>
      </div>
    </div>

    <div class="row" style="margin-top: 50px;">
      <div name="index" class="span3" style="position: fixed; width: 130px;">
	<ul class="nav nav-list">
	  <li class="nav-header">Resources</li>
<li><a href="#overview">overview</a></li><li><a href="#abundanceprofile">abundanceprofile</a></li><li><a href="#analysisset">analysisset</a></li><li><a href="#library">library</a></li><li><a href="#m5nr">m5nr</a></li><li><a href="#matrix">matrix</a></li><li><a href="#matrix">matrix</a></li><li><a href="#matrix_dump">matrix_dump</a></li><li><a href="#metadata">metadata</a></li><li><a href="#metagenome">metagenome</a></li><li><a href="#metagenome_statistics">metagenome_statistics</a></li><li><a href="#notebook">notebook</a></li><li><a href="#project">project</a></li><li><a href="#sample">sample</a></li><li><a href="#sequences">sequences</a></li><li><a href="#sequenceset">sequenceset</a></li></ul>	
    </div>
    <div class="span12" style="margin-left: 270px;">
<h1><a name='overview' style='padding-top: 50px;'>MG-RAST Overview</a></h1><p>RESTful Metagenomics RAST object and resource API
For usage note that required parameters need to be passed as path parameters, optional parameters need to be query parameters. If an optional parameter has a list of option values, the first displayed will be used as default.</p><hr><h1><a name='abundanceprofile' style='padding-top: 50px;'>abundanceprofile</a></h1><h2>Description</h2><p>A profile in biom format that contains abundance counts</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>instance - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>M5NR</b> - comprehensive protein database, type organism only</li><li><b>SwissProt</b> - protein database, type organism and feature only</li><li><b>GenBank</b> - protein database, type organism and feature only</li><li><b>IMG</b> - protein database, type organism and feature only</li><li><b>SEED</b> - protein database, type organism and feature only</li><li><b>TrEMBL</b> - protein database, type organism and feature only</li><li><b>RefSeq</b> - protein database, type organism and feature only</li><li><b>PATRIC</b> - protein database, type organism and feature only</li><li><b>KEGG</b> - protein database, type organism and feature only</li><li><b>M5RNA</b> - comprehensive RNA database, type organism only</li><li><b>RDP</b> - RNA database, type organism and feature only</li><li><b>Greengenes</b> - RNA database, type organism and feature only</li><li><b>LSU</b> - RNA database, type organism and feature only</li><li><b>SSU</b> - RNA database, type organism and feature only</li><li><b>Subsystems</b> - ontology database, type function only</li><li><b>NOG</b> - ontology database, type function only</li><li><b>COG</b> - ontology database, type function only</li><li><b>KO</b> - ontology database, type function only</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>length</b> (int)</li><p>value for minimum alignment length cutoff: default is 15</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>evalue</b> (int)</li><p>exponent value for maximum e-value cutoff: default is 5</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>identity</b> (int)</li><p>percent value for minimum % identity cutoff: default is 60</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>organism</b> - return organism data</li><li><b>function</b> - return functional data</li><li><b>feature</b> - return feature data</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>unique object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>generated_by</b> (string)</li><p>identifier of the data generator</p><li><b>date</b> (date)</li><p>time the output data was generated</p><li><b>matrix_type</b> (string)</li><p>type of the data encoding matrix (dense or sparse)</p><li><b>data</b> (list of lists of floats)</li><p>the matrix values</p><li><b>rows</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>matrix_element_type</b> (string)</li><p>data type of the elements in the return matrix</p><li><b>format_url</b> (string)</li><p>url to the format specification</p><li><b>format</b> (string)</li><p>format specification name</p><li><b>columns</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique identifier</li><li><b>metadata (hash)</b> - list of metadata, contains the metagenome</li></ul><br><li><b>shape</b> (list of integers)</li><p>list of the dimension sizes of the return matrix</p><li><b>type</b> (string)</li><p>type of the data in the return table (taxon, function or gene)</p><li><b>id</b> (string)</li><p>unique object identifier</p></p></ul><hr><h1><a name='analysisset' style='padding-top: 50px;'>analysisset</a></h1><h2>Description</h2><p>An analysis file from the processing of a metagenome from a specific stage in its analysis</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>instance - request</h2><h3>Description</h3><p>Returns a single sequence file.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>id</b> (string)</li><p>unique analysis set identifier - to get a list of all identifiers for a metagenome, use the setlist request</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>data</b> (file)</li><p>requested analysis file</p></p></ul><h2>setlist - request</h2><h3>Description</h3><p>Returns a list of sets of sequence files for the given id.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>id</b> (string)</li><p>unique metagenome identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>stage_name</b> (string)</li><p>name of the stage in processing of this file</p><li><b>file_name</b> (string)</li><p>name of the analysis file</p><li><b>url</b> (string)</li><p>url for retrieving this analysis file</p><li><b>stage_type</b> (string)</li><p>type of the analysis file within a stage, i.e. passed or removed for quality control steps</p><li><b>id</b> (string)</li><p>unique identifier of the analysis file</p><li><b>stage_id</b> (string)</li><p>three digit numerical identifier of the stage</p></p></ul><hr><h1><a name='library' style='padding-top: 50px;'>library</a></h1><h2>Description</h2><p>A library of metagenomic samples from some environment</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>query - request</h2><h3>Description</h3><p>Returns a set of data matching the query criteria.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>order</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>id</b> - return data objects ordered by id</li><li><b>name</b> - return data objects ordered by name</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>limit</b> (integer)</li><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</li><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>order</b> (string)</li><p>name of the attribute the returned data is ordered by</p><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>project (reference project)</b> - reference to the project object</li><li><b>version (integer)</b> - version of the object</li><li><b>name (string)</b> - human readable identifier</li><li><b>sequencesets (list of reference sequencesets)</b> - a list of references to the related sequence sets</li><li><b>metagenome (reference metagenome)</b> - reference to the related metagenome object</li><li><b>created (date)</b> - time the object was first created</li><li><b>url (uri)</b> - resource location of this object instance</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li><li><b>sample (reference sample)</b> - reference to the related sample object</li><li><b>id (string)</b> - unique object identifier</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p></p></ul><h2>instance - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li><li><b>verbose</b> - returns a standard subselection of metadata</li><li><b>full</b> - returns all connected metadata</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>unique object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>project</b> (reference project)</li><p>reference to the project object</p><li><b>version</b> (integer)</li><p>version of the object</p><li><b>name</b> (string)</li><p>human readable identifier</p><li><b>sequencesets</b> (list of reference sequencesets)</li><p>a list of references to the related sequence sets</p><li><b>metagenome</b> (reference metagenome)</li><p>reference to the related metagenome object</p><li><b>created</b> (date)</li><p>time the object was first created</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>metadata</b> (hash)</li><p>key value pairs describing metadata</p><li><b>sample</b> (reference sample)</li><p>reference to the related sample object</p><li><b>id</b> (string)</li><p>unique object identifier</p></p></ul><hr><h1><a name='m5nr' style='padding-top: 50px;'>m5nr</a></h1><h2>Description</h2><p>M5NR provides data through a comprehensive non-redundant protein / rRNA database</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>ontology - request</h2><h3>Description</h3><p></p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>Subsystems</b> - returns 4 level SEED-Subsystems ontology</li><li><b>COG</b> - returns 3 level COG ontology</li><li><b>NOG</b> - returns 3 level NOG ontology</li><li><b>KO</b> - returns 4 level KEGG-KO ontology</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>parent_name</b> (string)</li><p>name of ontology group to retrieve children of</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>min_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>function</b> - bottom ontology level</li><li><b>level3</b> - ontology level</li><li><b>level2</b> - ontology level</li><li><b>level1</b> - top ontology level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><p>This request returns a list, where each element is a list. The elements of this list contain requested ontology levels, from highest to lowest.</p><h2>taxonomy - request</h2><h3>Description</h3><p></p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>parent_name</b> (string)</li><p>name of taxanomy group to retrieve children of</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>min_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>species</b> - taxonomy level</li><li><b>genus</b> - taxonomy level</li><li><b>family</b> - taxonomy level</li><li><b>order</b> -  taxonomy level</li><li><b>class</b> -  taxonomy level</li><li><b>phylum</b> - taxonomy level</li><li><b>domain</b> - top taxonomy level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><p>This request returns a list, where each element is a list. The elements of this list contain requested taxonomy levels, from highest to lowest.</p><h2>sources - request</h2><h3>Description</h3><p></p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul></ul><h3>Return Attributes</h3><p>This request returns a list, where each element is a hash. They key of the hash is the source name, the value is a object.</p><hr><h1><a name='matrix' style='padding-top: 50px;'>matrix</a></h1><h2>Description</h2><p>A profile in biom format that contains abundance counts</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>organism - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>M5NR</b> - comprehensive protein database</li><li><b>RefSeq</b> - protein database</li><li><b>SwissProt</b> - protein database</li><li><b>GenBank</b> - protein database</li><li><b>IMG</b> - protein database</li><li><b>SEED</b> - protein database</li><li><b>TrEMBL</b> - protein database</li><li><b>PATRIC</b> - protein database</li><li><b>KEGG</b> - protein database</li><li><b>M5RNA</b> - comprehensive RNA database</li><li><b>RDP</b> - RNA database</li><li><b>Greengenes</b> - RNA database</li><li><b>LSU</b> - RNA database</li><li><b>SSU</b> - RNA database</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>result_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>abundance</b> - number of reads with hits in annotation</li><li><b>evalue</b> - average e-value exponent of hits in annotation</li><li><b>identity</b> - average percent identity of hits in annotation</li><li><b>length</b> - average alignment length of hits in annotation</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter</b> (string)</li><p>filter the return results to only include abundances based on genes with this function</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>group_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>strain</b> - bottom organism taxanomic level</li><li><b>species</b> - organism type level</li><li><b>genus</b> - organism taxanomic level</li><li><b>family</b> - organism taxanomic level</li><li><b>order</b> - organism taxanomic level</li><li><b>class</b> - organism taxanomic level</li><li><b>phylum</b> - organism taxanomic level</li><li><b>domain</b> - top organism taxanomic level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>length</b> (int)</li><p>value for minimum alignment length cutoff: default is 15</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>evalue</b> (int)</li><p>exponent value for maximum e-value cutoff: default is 5</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>identity</b> (int)</li><p>percent value for minimum % identity cutoff: default is 60</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>Subsystems</b> - ontology database, type function only</li><li><b>NOG</b> - ontology database, type function only</li><li><b>COG</b> - ontology database, type function only</li><li><b>KO</b> - ontology database, type function only</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>one or more metagenome or project unique identifier</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>hide_metadata</b> (boolean)</li><p>if false return metagenome metadata set in 'columns' object</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>generated_by</b> (string)</li><p>identifier of the data generator</p><li><b>date</b> (date)</li><p>time the output data was generated</p><li><b>matrix_type</b> (string)</li><p>type of the data encoding matrix (dense or sparse)</p><li><b>data</b> (list of lists of floats)</li><p>the matrix values</p><li><b>matrix_element_value</b> (string)</li><p>result_type of the elements in the return matrix</p><li><b>rows</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique annotation text</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>matrix_element_type</b> (string)</li><p>data type of the elements in the return matrix</p><li><b>format_url</b> (string)</li><p>url to the format specification</p><li><b>format</b> (string)</li><p>format specification name</p><li><b>columns</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique metagenome identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>shape</b> (list of integers)</li><p>list of the dimension sizes of the return matrix</p><li><b>type</b> (string)</li><p>type of the data in the return table (taxon, function or gene)</p><li><b>id</b> (string)</li><p>unique object identifier</p></p></ul><h2>function - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>Subsystems</b> - ontology database, type function only</li><li><b>NOG</b> - ontology database, type function only</li><li><b>COG</b> - ontology database, type function only</li><li><b>KO</b> - ontology database, type function only</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>result_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>abundance</b> - number of reads with hits in annotation</li><li><b>evalue</b> - average e-value exponent of hits in annotation</li><li><b>identity</b> - average percent identity of hits in annotation</li><li><b>length</b> - average alignment length of hits in annotation</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter</b> (string)</li><p>filter the return results to only include abundances based on genes with this organism</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>group_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>function</b> - bottom function ontology level</li><li><b>level3</b> - function ontology level</li><li><b>level2</b> - function ontology level</li><li><b>level1</b> - top function ontology level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>length</b> (int)</li><p>value for minimum alignment length cutoff: default is 15</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>evalue</b> (int)</li><p>exponent value for maximum e-value cutoff: default is 5</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>identity</b> (int)</li><p>percent value for minimum % identity cutoff: default is 60</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>M5NR</b> - comprehensive protein database</li><li><b>RefSeq</b> - protein database</li><li><b>SwissProt</b> - protein database</li><li><b>GenBank</b> - protein database</li><li><b>IMG</b> - protein database</li><li><b>SEED</b> - protein database</li><li><b>TrEMBL</b> - protein database</li><li><b>PATRIC</b> - protein database</li><li><b>KEGG</b> - protein database</li><li><b>M5RNA</b> - comprehensive RNA database</li><li><b>RDP</b> - RNA database</li><li><b>Greengenes</b> - RNA database</li><li><b>LSU</b> - RNA database</li><li><b>SSU</b> - RNA database</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>one or more metagenome or project unique identifier</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>hide_metadata</b> (boolean)</li><p>if false return metagenome metadata set in 'columns' object</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>generated_by</b> (string)</li><p>identifier of the data generator</p><li><b>date</b> (date)</li><p>time the output data was generated</p><li><b>matrix_type</b> (string)</li><p>type of the data encoding matrix (dense or sparse)</p><li><b>data</b> (list of lists of floats)</li><p>the matrix values</p><li><b>matrix_element_value</b> (string)</li><p>result_type of the elements in the return matrix</p><li><b>rows</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique annotation text</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>matrix_element_type</b> (string)</li><p>data type of the elements in the return matrix</p><li><b>format_url</b> (string)</li><p>url to the format specification</p><li><b>format</b> (string)</li><p>format specification name</p><li><b>columns</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique metagenome identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>shape</b> (list of integers)</li><p>list of the dimension sizes of the return matrix</p><li><b>type</b> (string)</li><p>type of the data in the return table (taxon, function or gene)</p><li><b>id</b> (string)</li><p>unique object identifier</p></p></ul><h2>feature - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>RefSeq</b> - protein database</li><li><b>SwissProt</b> - protein database</li><li><b>GenBank</b> - protein database</li><li><b>IMG</b> - protein database</li><li><b>SEED</b> - protein database</li><li><b>TrEMBL</b> - protein database</li><li><b>PATRIC</b> - protein database</li><li><b>KEGG</b> - protein database</li><li><b>RDP</b> - RNA database</li><li><b>Greengenes</b> - RNA database</li><li><b>LSU</b> - RNA database</li><li><b>SSU</b> - RNA database</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>result_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>abundance</b> - number of reads with hits in annotation</li><li><b>evalue</b> - average e-value exponent of hits in annotation</li><li><b>identity</b> - average percent identity of hits in annotation</li><li><b>length</b> - average alignment length of hits in annotation</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>one or more metagenome or project unique identifier</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>hide_metadata</b> (boolean)</li><p>if false return metagenome metadata set in 'columns' object</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>generated_by</b> (string)</li><p>identifier of the data generator</p><li><b>date</b> (date)</li><p>time the output data was generated</p><li><b>matrix_type</b> (string)</li><p>type of the data encoding matrix (dense or sparse)</p><li><b>data</b> (list of lists of floats)</li><p>the matrix values</p><li><b>matrix_element_value</b> (string)</li><p>result_type of the elements in the return matrix</p><li><b>rows</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique annotation text</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>matrix_element_type</b> (string)</li><p>data type of the elements in the return matrix</p><li><b>format_url</b> (string)</li><p>url to the format specification</p><li><b>format</b> (string)</li><p>format specification name</p><li><b>columns</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique metagenome identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>shape</b> (list of integers)</li><p>list of the dimension sizes of the return matrix</p><li><b>type</b> (string)</li><p>type of the data in the return table (taxon, function or gene)</p><li><b>id</b> (string)</li><p>unique object identifier</p></p></ul><hr><h1><a name='matrix' style='padding-top: 50px;'>matrix</a></h1><h2>Description</h2><p>A profile in biom format that contains abundance counts</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>organism - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>M5NR</b> - comprehensive protein database</li><li><b>RefSeq</b> - protein database</li><li><b>SwissProt</b> - protein database</li><li><b>GenBank</b> - protein database</li><li><b>IMG</b> - protein database</li><li><b>SEED</b> - protein database</li><li><b>TrEMBL</b> - protein database</li><li><b>PATRIC</b> - protein database</li><li><b>KEGG</b> - protein database</li><li><b>M5RNA</b> - comprehensive RNA database</li><li><b>RDP</b> - RNA database</li><li><b>Greengenes</b> - RNA database</li><li><b>LSU</b> - RNA database</li><li><b>SSU</b> - RNA database</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>result_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>abundance</b> - number of reads with hits in annotation</li><li><b>evalue</b> - average e-value exponent of hits in annotation</li><li><b>identity</b> - average percent identity of hits in annotation</li><li><b>length</b> - average alignment length of hits in annotation</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter</b> (string)</li><p>filter the return results to only include abundances based on genes with this function</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>group_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>strain</b> - bottom organism taxanomic level</li><li><b>species</b> - organism type level</li><li><b>genus</b> - organism taxanomic level</li><li><b>family</b> - organism taxanomic level</li><li><b>order</b> - organism taxanomic level</li><li><b>class</b> - organism taxanomic level</li><li><b>phylum</b> - organism taxanomic level</li><li><b>domain</b> - top organism taxanomic level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>length</b> (int)</li><p>value for minimum alignment length cutoff: default is 15</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>evalue</b> (int)</li><p>exponent value for maximum e-value cutoff: default is 5</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>identity</b> (int)</li><p>percent value for minimum % identity cutoff: default is 60</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>Subsystems</b> - ontology database, type function only</li><li><b>NOG</b> - ontology database, type function only</li><li><b>COG</b> - ontology database, type function only</li><li><b>KO</b> - ontology database, type function only</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>one or more metagenome or project unique identifier</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>hide_metadata</b> (boolean)</li><p>if false return metagenome metadata set in 'columns' object</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>generated_by</b> (string)</li><p>identifier of the data generator</p><li><b>date</b> (date)</li><p>time the output data was generated</p><li><b>matrix_type</b> (string)</li><p>type of the data encoding matrix (dense or sparse)</p><li><b>data</b> (list of lists of floats)</li><p>the matrix values</p><li><b>matrix_element_value</b> (string)</li><p>result_type of the elements in the return matrix</p><li><b>rows</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique annotation text</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>matrix_element_type</b> (string)</li><p>data type of the elements in the return matrix</p><li><b>format_url</b> (string)</li><p>url to the format specification</p><li><b>format</b> (string)</li><p>format specification name</p><li><b>columns</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique metagenome identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>shape</b> (list of integers)</li><p>list of the dimension sizes of the return matrix</p><li><b>type</b> (string)</li><p>type of the data in the return table (taxon, function or gene)</p><li><b>id</b> (string)</li><p>unique object identifier</p></p></ul><h2>function - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>Subsystems</b> - ontology database, type function only</li><li><b>NOG</b> - ontology database, type function only</li><li><b>COG</b> - ontology database, type function only</li><li><b>KO</b> - ontology database, type function only</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>result_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>abundance</b> - number of reads with hits in annotation</li><li><b>evalue</b> - average e-value exponent of hits in annotation</li><li><b>identity</b> - average percent identity of hits in annotation</li><li><b>length</b> - average alignment length of hits in annotation</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter</b> (string)</li><p>filter the return results to only include abundances based on genes with this organism</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>group_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>function</b> - bottom function ontology level</li><li><b>level3</b> - function ontology level</li><li><b>level2</b> - function ontology level</li><li><b>level1</b> - top function ontology level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>length</b> (int)</li><p>value for minimum alignment length cutoff: default is 15</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>evalue</b> (int)</li><p>exponent value for maximum e-value cutoff: default is 5</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>identity</b> (int)</li><p>percent value for minimum % identity cutoff: default is 60</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>M5NR</b> - comprehensive protein database</li><li><b>RefSeq</b> - protein database</li><li><b>SwissProt</b> - protein database</li><li><b>GenBank</b> - protein database</li><li><b>IMG</b> - protein database</li><li><b>SEED</b> - protein database</li><li><b>TrEMBL</b> - protein database</li><li><b>PATRIC</b> - protein database</li><li><b>KEGG</b> - protein database</li><li><b>M5RNA</b> - comprehensive RNA database</li><li><b>RDP</b> - RNA database</li><li><b>Greengenes</b> - RNA database</li><li><b>LSU</b> - RNA database</li><li><b>SSU</b> - RNA database</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>one or more metagenome or project unique identifier</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>hide_metadata</b> (boolean)</li><p>if false return metagenome metadata set in 'columns' object</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>generated_by</b> (string)</li><p>identifier of the data generator</p><li><b>date</b> (date)</li><p>time the output data was generated</p><li><b>matrix_type</b> (string)</li><p>type of the data encoding matrix (dense or sparse)</p><li><b>data</b> (list of lists of floats)</li><p>the matrix values</p><li><b>matrix_element_value</b> (string)</li><p>result_type of the elements in the return matrix</p><li><b>rows</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique annotation text</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>matrix_element_type</b> (string)</li><p>data type of the elements in the return matrix</p><li><b>format_url</b> (string)</li><p>url to the format specification</p><li><b>format</b> (string)</li><p>format specification name</p><li><b>columns</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique metagenome identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>shape</b> (list of integers)</li><p>list of the dimension sizes of the return matrix</p><li><b>type</b> (string)</li><p>type of the data in the return table (taxon, function or gene)</p><li><b>id</b> (string)</li><p>unique object identifier</p></p></ul><h2>feature - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>RefSeq</b> - protein database</li><li><b>SwissProt</b> - protein database</li><li><b>GenBank</b> - protein database</li><li><b>IMG</b> - protein database</li><li><b>SEED</b> - protein database</li><li><b>TrEMBL</b> - protein database</li><li><b>PATRIC</b> - protein database</li><li><b>KEGG</b> - protein database</li><li><b>RDP</b> - RNA database</li><li><b>Greengenes</b> - RNA database</li><li><b>LSU</b> - RNA database</li><li><b>SSU</b> - RNA database</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>result_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>abundance</b> - number of reads with hits in annotation</li><li><b>evalue</b> - average e-value exponent of hits in annotation</li><li><b>identity</b> - average percent identity of hits in annotation</li><li><b>length</b> - average alignment length of hits in annotation</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>one or more metagenome or project unique identifier</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>hide_metadata</b> (boolean)</li><p>if false return metagenome metadata set in 'columns' object</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>generated_by</b> (string)</li><p>identifier of the data generator</p><li><b>date</b> (date)</li><p>time the output data was generated</p><li><b>matrix_type</b> (string)</li><p>type of the data encoding matrix (dense or sparse)</p><li><b>data</b> (list of lists of floats)</li><p>the matrix values</p><li><b>matrix_element_value</b> (string)</li><p>result_type of the elements in the return matrix</p><li><b>rows</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique annotation text</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>matrix_element_type</b> (string)</li><p>data type of the elements in the return matrix</p><li><b>format_url</b> (string)</li><p>url to the format specification</p><li><b>format</b> (string)</li><p>format specification name</p><li><b>columns</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique metagenome identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>shape</b> (list of integers)</li><p>list of the dimension sizes of the return matrix</p><li><b>type</b> (string)</li><p>type of the data in the return table (taxon, function or gene)</p><li><b>id</b> (string)</li><p>unique object identifier</p></p></ul><hr><h1><a name='matrix_dump' style='padding-top: 50px;'>matrix_dump</a></h1><h2>Description</h2><p>A profile in biom format that contains abundance counts</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>organism - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>M5NR</b> - comprehensive protein database</li><li><b>RefSeq</b> - protein database</li><li><b>SwissProt</b> - protein database</li><li><b>GenBank</b> - protein database</li><li><b>IMG</b> - protein database</li><li><b>SEED</b> - protein database</li><li><b>TrEMBL</b> - protein database</li><li><b>PATRIC</b> - protein database</li><li><b>KEGG</b> - protein database</li><li><b>M5RNA</b> - comprehensive RNA database</li><li><b>RDP</b> - RNA database</li><li><b>Greengenes</b> - RNA database</li><li><b>LSU</b> - RNA database</li><li><b>SSU</b> - RNA database</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>result_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>abundance</b> - number of reads with hits in annotation</li><li><b>evalue</b> - average e-value exponent of hits in annotation</li><li><b>identity</b> - average percent identity of hits in annotation</li><li><b>length</b> - average alignment length of hits in annotation</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter</b> (string)</li><p>filter the return results to only include abundances based on genes with this function</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>group_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>strain</b> - bottom organism taxanomic level</li><li><b>species</b> - organism type level</li><li><b>genus</b> - organism taxanomic level</li><li><b>family</b> - organism taxanomic level</li><li><b>order</b> - organism taxanomic level</li><li><b>class</b> - organism taxanomic level</li><li><b>phylum</b> - organism taxanomic level</li><li><b>domain</b> - top organism taxanomic level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>length</b> (int)</li><p>value for minimum alignment length cutoff: default is 15</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>evalue</b> (int)</li><p>exponent value for maximum e-value cutoff: default is 5</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>identity</b> (int)</li><p>percent value for minimum % identity cutoff: default is 60</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>Subsystems</b> - ontology database, type function only</li><li><b>NOG</b> - ontology database, type function only</li><li><b>COG</b> - ontology database, type function only</li><li><b>KO</b> - ontology database, type function only</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>one or more metagenome or project unique identifier</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>generated_by</b> (string)</li><p>identifier of the data generator</p><li><b>date</b> (date)</li><p>time the output data was generated</p><li><b>matrix_type</b> (string)</li><p>type of the data encoding matrix (dense or sparse)</p><li><b>data</b> (list of lists of floats)</li><p>the matrix values</p><li><b>matrix_element_value</b> (string)</li><p>result_type of the elements in the return matrix</p><li><b>rows</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique annotation text</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>matrix_element_type</b> (string)</li><p>data type of the elements in the return matrix</p><li><b>format_url</b> (string)</li><p>url to the format specification</p><li><b>format</b> (string)</li><p>format specification name</p><li><b>columns</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique metagenome identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>shape</b> (list of integers)</li><p>list of the dimension sizes of the return matrix</p><li><b>type</b> (string)</li><p>type of the data in the return table (taxon, function or gene)</p><li><b>id</b> (string)</li><p>unique object identifier</p></p></ul><h2>function - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>Subsystems</b> - ontology database, type function only</li><li><b>NOG</b> - ontology database, type function only</li><li><b>COG</b> - ontology database, type function only</li><li><b>KO</b> - ontology database, type function only</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>result_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>abundance</b> - number of reads with hits in annotation</li><li><b>evalue</b> - average e-value exponent of hits in annotation</li><li><b>identity</b> - average percent identity of hits in annotation</li><li><b>length</b> - average alignment length of hits in annotation</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter</b> (string)</li><p>filter the return results to only include abundances based on genes with this organism</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>group_level</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>function</b> - bottom function ontology level</li><li><b>level3</b> - function ontology level</li><li><b>level2</b> - function ontology level</li><li><b>level1</b> - top function ontology level</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>length</b> (int)</li><p>value for minimum alignment length cutoff: default is 15</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>evalue</b> (int)</li><p>exponent value for maximum e-value cutoff: default is 5</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>identity</b> (int)</li><p>percent value for minimum % identity cutoff: default is 60</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>filter_source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>M5NR</b> - comprehensive protein database</li><li><b>RefSeq</b> - protein database</li><li><b>SwissProt</b> - protein database</li><li><b>GenBank</b> - protein database</li><li><b>IMG</b> - protein database</li><li><b>SEED</b> - protein database</li><li><b>TrEMBL</b> - protein database</li><li><b>PATRIC</b> - protein database</li><li><b>KEGG</b> - protein database</li><li><b>M5RNA</b> - comprehensive RNA database</li><li><b>RDP</b> - RNA database</li><li><b>Greengenes</b> - RNA database</li><li><b>LSU</b> - RNA database</li><li><b>SSU</b> - RNA database</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>one or more metagenome or project unique identifier</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>generated_by</b> (string)</li><p>identifier of the data generator</p><li><b>date</b> (date)</li><p>time the output data was generated</p><li><b>matrix_type</b> (string)</li><p>type of the data encoding matrix (dense or sparse)</p><li><b>data</b> (list of lists of floats)</li><p>the matrix values</p><li><b>matrix_element_value</b> (string)</li><p>result_type of the elements in the return matrix</p><li><b>rows</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique annotation text</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>matrix_element_type</b> (string)</li><p>data type of the elements in the return matrix</p><li><b>format_url</b> (string)</li><p>url to the format specification</p><li><b>format</b> (string)</li><p>format specification name</p><li><b>columns</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique metagenome identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>shape</b> (list of integers)</li><p>list of the dimension sizes of the return matrix</p><li><b>type</b> (string)</li><p>type of the data in the return table (taxon, function or gene)</p><li><b>id</b> (string)</li><p>unique object identifier</p></p></ul><h2>feature - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>RefSeq</b> - protein database</li><li><b>SwissProt</b> - protein database</li><li><b>GenBank</b> - protein database</li><li><b>IMG</b> - protein database</li><li><b>SEED</b> - protein database</li><li><b>TrEMBL</b> - protein database</li><li><b>PATRIC</b> - protein database</li><li><b>KEGG</b> - protein database</li><li><b>RDP</b> - RNA database</li><li><b>Greengenes</b> - RNA database</li><li><b>LSU</b> - RNA database</li><li><b>SSU</b> - RNA database</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>result_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>abundance</b> - number of reads with hits in annotation</li><li><b>evalue</b> - average e-value exponent of hits in annotation</li><li><b>identity</b> - average percent identity of hits in annotation</li><li><b>length</b> - average alignment length of hits in annotation</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>one or more metagenome or project unique identifier</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>generated_by</b> (string)</li><p>identifier of the data generator</p><li><b>date</b> (date)</li><p>time the output data was generated</p><li><b>matrix_type</b> (string)</li><p>type of the data encoding matrix (dense or sparse)</p><li><b>data</b> (list of lists of floats)</li><p>the matrix values</p><li><b>matrix_element_value</b> (string)</li><p>result_type of the elements in the return matrix</p><li><b>rows</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique annotation text</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>matrix_element_type</b> (string)</li><p>data type of the elements in the return matrix</p><li><b>format_url</b> (string)</li><p>url to the format specification</p><li><b>format</b> (string)</li><p>format specification name</p><li><b>columns</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>id (string)</b> - unique metagenome identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>shape</b> (list of integers)</li><p>list of the dimension sizes of the return matrix</p><li><b>type</b> (string)</li><p>type of the data in the return table (taxon, function or gene)</p><li><b>id</b> (string)</li><p>unique object identifier</p></p></ul><hr><h1><a name='metadata' style='padding-top: 50px;'>metadata</a></h1><h2>Description</h2><p>Metagenomic metadata is data providing information about one or more aspects of a set sequences from a sample of some environment</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>template - request</h2><h3>Description</h3><p></p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul></ul><h3>Return Attributes</h3><ul><li><b>ep</b> (hash of keys of strings)</li><p>enviromental package type</p><li><b>project</b> (hash of keys of strings)</li><p>project type</p><li><b>sample</b> (hash of keys of strings)</li><p>sample type</p><li><b>library</b> (hash of keys of strings)</li><p>library type</p></p></ul><h2>cv - request</h2><h3>Description</h3><p></p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul></ul><h3>Return Attributes</h3><ul><li><b>ontology</b> (hash of keys of strings)</li><p>metadata label</p><li><b>ont_id</b> (hash of keys of strings)</li><p>metadata label</p><li><b>select</b> (hash of keys of strings)</li><p>metadata label</p></p></ul><h2>export - request</h2><h3>Description</h3><p></p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>id</b> (string)</li><p>unique object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>samples</b> ()</li><p></p><li><b>sampleNum</b> ()</li><p></p><li><b>name</b> ()</li><p></p><li><b>id</b> ()</li><p></p><li><b>data</b> ()</li><p></p></p></ul><h2>validate - request</h2><h3>Description</h3><p></p><p>This is a synchronous POST request.</p><h3>Parameters</h3><ul><li><b>metadata_spreadsheet</b> (file)</p><p>xlsx or xls format spreadsheet with metadata</p><p>This parameter must be passed in the message body.</p></ul><h3>Return Attributes</h3><ul><li><b>metadata</b> (object)</li><p>valid metadata object for project and its samples and libraries</p><li><b>is_valid</b> (boolean)</li><p>the inputed value is valid for the given category and label</p></p></ul><h2>validate - request</h2><h3>Description</h3><p></p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>group</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>mixs</b> - label is part of MIxS (minimal) metadata</li><li><b>mims</b> - label is part of MIMS (metagenome) metadata</li><li><b>migs</b> - label is part of MIGS (genome) metadata</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>value</b> (string)</li><p>metadata value</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>category</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>project</b> - label belongs to project metadata</li><li><b>sample</b> - label belongs to sample metadata</li><li><b>library</b> - label belongs to library metadata</li><li><b>env_package</b> - label belongs to env_package metadata</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>label</b> (string)</li><p>metadata label</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>is_valid</b> (boolean)</li><p>the inputed value is valid for the given category and label</p><li><b>message</b> (string)</li><p>if not valid, reason why</p></p></ul><hr><h1><a name='metagenome' style='padding-top: 50px;'>metagenome</a></h1><h2>Description</h2><p>A metagenome is an analyzed set sequences from a sample of some environment</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>query - request</h2><h3>Description</h3><p>Returns a set of data matching the query criteria.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li><li><b>verbose</b> - returns a standard subselection of metadata</li><li><b>full</b> - returns all connected metadata</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>order</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>id</b> - return data objects ordered by id</li><li><b>name</b> - return data objects ordered by name</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>limit</b> (integer)</li><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</li><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>order</b> (string)</li><p>name of the attribute the returned data is ordered by</p><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>project (reference project)</b> - reference to the project object</li><li><b>version (integer)</b> - version of the object</li><li><b>status (cv of ARRAY(0x7feae92fe918)s of privates)</b> - object is private</li><li><b>name (string)</b> - human readable identifier</li><li><b>library (reference library)</b> - reference to the related library object</li><li><b>sequence_type (string)</b> - sequencing type</li><li><b>created (date)</b> - time the object was first created</li><li><b>url (uri)</b> - resource location of this object instance</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li><li><b>sample (reference sample)</b> - reference to the related sample object</li><li><b>id (string)</b> - unique object identifier</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p></p></ul><h2>instance - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li><li><b>verbose</b> - returns a standard subselection of metadata</li><li><b>full</b> - returns all connected metadata</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>unique object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>project</b> (reference project)</li><p>reference to the project object</p><li><b>version</b> (integer)</li><p>version of the object</p><li><b>status</b> (cv of ARRAY(0x7feae92eb100)s of privates)</li><p>object is private</p><li><b>name</b> (string)</li><p>human readable identifier</p><li><b>library</b> (reference library)</li><p>reference to the related library object</p><li><b>sequence_type</b> (string)</li><p>sequencing type</p><li><b>created</b> (date)</li><p>time the object was first created</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>metadata</b> (hash)</li><p>key value pairs describing metadata</p><li><b>sample</b> (reference sample)</li><p>reference to the related sample object</p><li><b>id</b> (string)</li><p>unique object identifier</p></p></ul><hr><h1><a name='metagenome_statistics' style='padding-top: 50px;'>metagenome_statistics</a></h1><h2>Description</h2><p>An set of statistical information obtained during the analysis of a metagenomic sequence</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>instance - request</h2><h3>Description</h3><p>Returns a JSON structure of statistical information.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only sequence_stats attribute</li><li><b>verbose</b> - returns all but qc attribute</li><li><b>full</b> - returns all attributes</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>metagenome id</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>source</b> (hash)</li><p>evalue and % identity counts per source</p><li><b>rarefaction</b> (list)</li><p>rarefaction coordinate data</p><li><b>sequence_stats</b> (hash)</li><p>statistics on sequence files of all pipeline stages</p><li><b>id</b> (string)</li><p>unique metagenome id</p></p></ul><hr><h1><a name='notebook' style='padding-top: 50px;'>notebook</a></h1><h2>Description</h2><p>A notebook is a JSON structure describing the contents of an ipython session.</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>query - request</h2><h3>Description</h3><p>Returns a set of data matching the query criteria.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns notebook attributes</li><li><b>full</b> - returns notebook attributes and object</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>order</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>id</b> - return data objects ordered by id</li><li><b>name</b> - return data objects ordered by name</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>limit</b> (integer)</li><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</li><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>order</b> (string)</li><p>name of the attribute the returned data is ordered by</p><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>version (integer)</b> - version of the object</li><li><b>status (cv of ARRAY(0x7feae9301690)s of privates)</b> - notebook is private</li><li><b>name (string)</b> - human readable identifier</li><li><b>permission (cv of ARRAY(0x7feae9301828)s of edits)</b> - notebook is editable</li><li><b>notebook (object)</b> - notebook object in JSON format</li><li><b>uuid (string)</b> - ipynb identifier - stable across different versions</li><li><b>created (date)</b> - time the object was first created</li><li><b>url (uri)</b> - resource location of this object instance</li><li><b>id (string)</b> - unique shock identifier</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p></p></ul><h2>instance - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns notebook attributes</li><li><b>full</b> - returns notebook attributes and object</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>unique shock object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>version</b> (integer)</li><p>version of the object</p><li><b>status</b> (cv of ARRAY(0x7feae9305830)s of privates)</li><p>notebook is private</p><li><b>name</b> (string)</li><p>human readable identifier</p><li><b>permission</b> (cv of ARRAY(0x7feae93059c8)s of edits)</li><p>notebook is editable</p><li><b>notebook</b> (object)</li><p>notebook object in JSON format</p><li><b>uuid</b> (string)</li><p>ipynb identifier - stable across different versions</p><li><b>created</b> (date)</li><p>time the object was first created</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>id</b> (string)</li><p>unique shock identifier</p></p></ul><h2>clone - request</h2><h3>Description</h3><p>Clones a data object with new notebook id and returns it.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns notebook attributes</li><li><b>full</b> - returns notebook attributes and object</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>name</b> (string)</li><p>name of new cloned notebook</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>nbid</b> (string)</li><p>unique notebook object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p><li><b>id</b> (string)</li><p>unique shock object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>version</b> (integer)</li><p>version of the object</p><li><b>status</b> (cv of ARRAY(0x7feae9306130)s of privates)</li><p>notebook is private</p><li><b>name</b> (string)</li><p>human readable identifier</p><li><b>permission</b> (cv of ARRAY(0x7feae93062c8)s of edits)</li><p>notebook is editable</p><li><b>notebook</b> (object)</li><p>notebook object in JSON format</p><li><b>uuid</b> (string)</li><p>ipynb identifier - stable across different versions</p><li><b>created</b> (date)</li><p>time the object was first created</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>id</b> (string)</li><p>unique shock identifier</p></p></ul><h2>upload - request</h2><h3>Description</h3><p>Upload a notebook to shock.</p><p>This is a synchronous POST request.</p><h3>Parameters</h3><ul><li><b>ipynb</b> (file)</p><p>.pynb file in JSON format</p><p>This parameter must be passed in the message body.</p></ul><h3>Return Attributes</h3><ul><li><b>version</b> (integer)</li><p>version of the object</p><li><b>status</b> (cv of ARRAY(0x7feae9306860)s of privates)</li><p>notebook is private</p><li><b>name</b> (string)</li><p>human readable identifier</p><li><b>permission</b> (cv of ARRAY(0x7feae93069f8)s of edits)</li><p>notebook is editable</p><li><b>notebook</b> (object)</li><p>notebook object in JSON format</p><li><b>uuid</b> (string)</li><p>ipynb identifier - stable across different versions</p><li><b>created</b> (date)</li><p>time the object was first created</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>id</b> (string)</li><p>unique shock identifier</p></p></ul><hr><h1><a name='project' style='padding-top: 50px;'>project</a></h1><h2>Description</h2><p>A project is a composition of samples, libraries and metagenomes being analyzed in a global context.</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>query - request</h2><h3>Description</h3><p>Returns a set of data matching the query criteria.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li><li><b>verbose</b> - returns all metadata</li><li><b>full</b> - returns all metadata and references</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>order</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>id</b> - return data objects ordered by id</li><li><b>name</b> - return data objects ordered by name</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>limit</b> (integer)</li><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</li><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>order</b> (string)</li><p>name of the attribute the returned data is ordered by</p><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>analyzed (list of reference metagenomes)</b> - a list of references to the related metagenome objects</li><li><b>version (integer)</b> - version of the object</li><li><b>status (cv of ARRAY(0x7feae93077a8)s of privates)</b> - object is private</li><li><b>name (string)</b> - human readable identifier</li><li><b>libraries (list of reference librarys)</b> - a list of references to the related library objects</li><li><b>description (string)</b> - a short, comprehensive description of the project</li><li><b>created (date)</b> - time the object was first created</li><li><b>samples (list of reference samples)</b> - a list of references to the related sample objects</li><li><b>funding_source (string)</b> - the official name of the source of funding of this project</li><li><b>url (uri)</b> - resource location of this object instance</li><li><b>pi (string)</b> - the first and last name of the principal investigator of the project</li><li><b>id (string)</b> - unique object identifier</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p></p></ul><h2>instance - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li><li><b>verbose</b> - returns all metadata</li><li><b>full</b> - returns all metadata and references</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>unique object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>analyzed</b> (list of reference metagenomes)</li><p>a list of references to the related metagenome objects</p><li><b>version</b> (integer)</li><p>version of the object</p><li><b>status</b> (cv of ARRAY(0x7feae930a328)s of privates)</li><p>object is private</p><li><b>name</b> (string)</li><p>human readable identifier</p><li><b>libraries</b> (list of reference librarys)</li><p>a list of references to the related library objects</p><li><b>description</b> (string)</li><p>a short, comprehensive description of the project</p><li><b>created</b> (date)</li><p>time the object was first created</p><li><b>samples</b> (list of reference samples)</li><p>a list of references to the related sample objects</p><li><b>funding_source</b> (string)</li><p>the official name of the source of funding of this project</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>pi</b> (string)</li><p>the first and last name of the principal investigator of the project</p><li><b>id</b> (string)</li><p>unique object identifier</p><li><b>metadata</b> (hash)</li><p>key value pairs describing metadata</p></p></ul><hr><h1><a name='sample' style='padding-top: 50px;'>sample</a></h1><h2>Description</h2><p>A metagenomic sample from some environment.</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>query - request</h2><h3>Description</h3><p>Returns a set of data matching the query criteria.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>order</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>id</b> - return data objects ordered by id</li><li><b>name</b> - return data objects ordered by name</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>limit</b> (integer)</li><p>maximum number of items requested</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>offset</b> (integer)</li><p>zero based index of the first data object to be returned</p><p>This is an optional parameter and may be passed in the query string.</p></ul><h3>Return Attributes</h3><ul><li><b>next</b> (uri)</li><p>link to the previous set or null if this is the first set</p><li><b>prev</b> (uri)</li><p>link to the next set or null if this is the last set</p><li><b>order</b> (string)</li><p>name of the attribute the returned data is ordered by</p><li><b>data</b> (list of objects)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>project (reference project)</b> - reference to the project of this sample</li><li><b>version (integer)</b> - version of the object</li><li><b>name (string)</b> - human readable identifier</li><li><b>libraries (list of reference librarys)</b> - a list of references to the related library objects</li><li><b>metagenomes (list of reference metagenomes)</b> - a list of references to the related metagenome objects</li><li><b>created (date)</b> - time the object was first created</li><li><b>env_package (object of HASH(0x7feae930d400)s)</b> - environmental package object</li><li><b>url (uri)</b> - resource location of this object instance</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li><li><b>id (string)</b> - unique object identifier</li></ul><br><li><b>limit</b> (integer)</li><p>maximum number of data items returned, default is 10</p><li><b>total_count</b> (integer)</li><p>total number of available data items</p><li><b>offset</b> (integer)</li><p>zero based index of the first returned data item</p></p></ul><h2>instance - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>verbosity</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>minimal</b> - returns only minimal information</li><li><b>verbose</b> - returns all metadata</li><li><b>full</b> - returns all metadata and references</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>unique object identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>project</b> (reference project)</li><p>reference to the project of this sample</p><li><b>version</b> (integer)</li><p>version of the object</p><li><b>name</b> (string)</li><p>human readable identifier</p><li><b>libraries</b> (list of reference librarys)</li><p>a list of references to the related library objects</p><li><b>metagenomes</b> (list of reference metagenomes)</li><p>a list of references to the related metagenome objects</p><li><b>created</b> (date)</li><p>time the object was first created</p><li><b>env_package</b> (object)</li><p>This attribute has an object structure:</p><ul style='list-style: none;'><li><b>created (date)</b> - creation date</li><li><b>name (string)</b> - name of the package</li><li><b>metadata (hash)</b> - key value pairs describing metadata</li><li><b>type (string)</b> - package type</li><li><b>id (string)</b> - unique package identifier</li></ul><br><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>metadata</b> (hash)</li><p>key value pairs describing metadata</p><li><b>id</b> (string)</li><p>unique object identifier</p></p></ul><hr><h1><a name='sequences' style='padding-top: 50px;'>sequences</a></h1><h2>Description</h2><p>A set of genomic sequences of a metagenome annotated by a specified source</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>md5 - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>sequence_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>dna</b> - return DNA sequences</li><li><b>protein</b> - return protein sequences</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>md5</b> (list of string)</li><p>md5 identifier</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>unique metagenome identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>version</b> (integer)</li><p>version of the object</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>id</b> (string)</li><p>unique object identifier</p><li><b>data</b> (list of hashs)</li><p>a hash of data_type to list of sequences</p></p></ul><h2>annotation - request</h2><h3>Description</h3><p>Returns a single data object.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>source</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>RDP</b> - RNA database, type organism and feature only</li><li><b>Greengenes</b> - RNA database, type organism and feature only</li><li><b>LSU</b> - RNA database, type organism and feature only</li><li><b>SSU</b> - RNA database, type organism and feature only</li><li><b>SwissProt</b> - protein database, type organism and feature only</li><li><b>GenBank</b> - protein database, type organism and feature only</li><li><b>IMG</b> - protein database, type organism and feature only</li><li><b>SEED</b> - protein database, type organism and feature only</li><li><b>TrEMBL</b> - protein database, type organism and feature only</li><li><b>RefSeq</b> - protein database, type organism and feature only</li><li><b>PATRIC</b> - protein database, type organism and feature only</li><li><b>eggNOG</b> - protein database, type organism and feature only</li><li><b>KEGG</b> - protein database, type organism and feature only</li><li><b>NOG</b> - ontology database, type function only</li><li><b>COG</b> - ontology database, type function only</li><li><b>KO</b> - ontology database, type function only</li><li><b>GO</b> - ontology database, type function only</li><li><b>Subsystems</b> - ontology database, type function only</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>data_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>organism</b> - return organism data</li><li><b>function</b> - return function data</li><li><b>ontology</b> - return ontology data</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>sequence_type</b> (controlled vocabulary)</li><p>This parameter value can be chosen from the following (the first being default):</p><ul style='list-style: none;'><li><b>dna</b> - return DNA sequences</li><li><b>protein</b> - return protein sequences</li></ul><br><p>This is an optional parameter and may be passed in the query string.</p><li><b>annotation</b> (list of string)</li><p>data_type to filter by</p><p>This is an optional parameter and may be passed in the query string.</p><li><b>id</b> (string)</li><p>unique metagenome identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>version</b> (integer)</li><p>version of the object</p><li><b>url</b> (uri)</li><p>resource location of this object instance</p><li><b>id</b> (string)</li><p>unique object identifier</p><li><b>data</b> (list of hashs)</li><p>a hash of data_type to list of sequences</p></p></ul><hr><h1><a name='sequenceset' style='padding-top: 50px;'>sequenceset</a></h1><h2>Description</h2><p>A set / subset of genomic sequences of a metagenome from a specific stage in its analysis</p><h2>info - request</h2><h3>Description</h3><p>Returns description of parameters and attributes.</p><p>This is a synchronous GET request. This request has no parameters.<h2>instance - request</h2><h3>Description</h3><p>Returns a single sequence file.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>id</b> (string)</li><p>unique sequence set identifier - to get a list of all identifiers for a metagenome, use the setlist request</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>data</b> (file)</li><p>requested sequence file</p></p></ul><h2>setlist - request</h2><h3>Description</h3><p>Returns a list of sets of sequence files for the given id.</p><p>This is a synchronous GET request.</p><h3>Parameters</h3><ul><li><b>id</b> (string)</li><p>unique metagenome identifier</p><p>This is a required parameter and must be passed as a REST parameter.</p></ul><h3>Return Attributes</h3><ul><li><b>stage_name</b> (string)</li><p>name of the stage in processing of this sequence file</p><li><b>file_name</b> (string)</li><p>name of the sequence file</p><li><b>url</b> (string)</li><p>url for retrieving this sequence file</p><li><b>stage_type</b> (string)</li><p>type of the sequence file within a stage, i.e. passed or removed for quality control steps</p><li><b>id</b> (string)</li><p>unique identifier of the sequence file</p><li><b>stage_id</b> (string)</li><p>three digit numerical identifier of the stage</p></p></ul><hr>
</ul>
    </div>
  </body>

</html>
