package MGRAST::WebPage::About;

use base qw( WebPage );

use strict;
use warnings;

use WebConfig;
use FIG_Config;

1;

=pod

=head1 NAME

MetagenomeSelect - an instance of WebPage which lets the user select a metagenome

=head1 DESCRIPTION

Display an metagenome select box

=head1 METHODS

=over 4

=item * B<init> ()

Called when the web page is instanciated.

=cut

sub init {
  my ($self) = @_;

  $self->title('About');
  $self->application->register_component('TabView', 'tv');
  $self->data('dataset_select_metabolic', [ 'set1' ]);
  
  return 1;
}


=pod

=item * B<output> ()

Returns the html output of the OrganismSelect page.

=cut

sub output {
  my ($self) = @_;

  my $application = $self->application;
  my $tv = $application->component('tv');
  $tv->width(750);
  my $cgi = $application->cgi;

  my $metagenome = $cgi->param('metagenome') || "";

  my $description = "<p>Researchers can upload their sample's metagenome, and MG-RAST uses a variety of computing resources - Argonne's 800-core cluster, TeraGrid and cloud computing - to compare the DNA fragments to those from every other sample in the system as well as to gene sequences in several other publicly-available databases.</p>";
  
  $description .= '<p>The service is built as a modified version of the <a href="http://rast.nmpdr.org" target="_blank">RAST server</a>, which was originally designed to support high-quality annotation of complete or draft microbial genomes. We have adapted this technology for the analysis of metagenomes and are currently able to handle fasta files generated by pyrosequencing (454) and Sanger sequencing data. In addition to <a href="http://www.theseed.org" target="_blank">SEED</a> and <a href="http://www.nmpdr.org" target=_blank><img src="'.$FIG_Config::cgi_url.'/Html/nmpdr_icon_small.png">&nbsp;NMPDR</a> data we use the following ribosomal RNA databases for our analyses: <a href="http://greengenes.lbl.gov" target="_blank">GREENGENES</a>, <a href="http://rdp.cme.msu.edu/" target="_blank">RDP-II</a>, <a href="http://www.arb-silva.de/" target="_blank">SILVA</a>, and <a href="http://bioinformatics.psb.ugent.be/webtools/rRNA/" target="_blank">European ribosomal RNA database.</a></p></p>';
  
  $description .= '<p>User submission and analysis are confidential, although each user may optionally allow a collaborative environment that enables multiple users to share datasets. In addition, users can also allow public access and request long-term storage of their metagenomic samples on the server. The server provides unique IDs for metagenomes and the sequences they contain and provides a stable mechanism for linking to. All data and analysis are available for download in a variety of different formats.</p>';
  
  $description .= "<p>To be able to contact you once the computation is finished and in case user intervention is required, we request that users register with a valid email address.</p>\n";

  $description .= "<p>We request that users register with a valid email address so we can contact you once the computation is finished or in case user intervention is required.</p>\n";

  my $intro ="<h3>About MG-RAST</h3><p>The MG-RAST server (MetaGenome Rapid Annotation using Subsystems Technology) is a fully automated analysis system that has incorporated a variety of tools to help users explore datasets via an intuitive web-based platform. The system has found widespread use in the analysis of metagenome-derived datasets, as the data exploration can be performed in 'real-time' and comparative analysis can be made against other metagenomes and complete genomes. Currently MG-RAST has processed over 4300 data sets (>400 public) and has over 1500 users.</p>";

  $intro .="<p align=\"right\"><a onclick=\"switch_display('description')\" style='cursor:pointer;color:blue'><table><tr><td valign=\"center\"'><img id='expand_img' src=\"$FIG_Config::cgi_url/Html/MGRAST-more.png\" onclick=\"if (document.getElementById('description').style.display == 'none') { document.getElementById('description').style.display = 'inline'; } else { document.getElementById('description').style.display = 'none'; };\"></td><td style=\"vertical-align:middle;\" id='expand_text'>more</td></tr></table></a></p>";

  my $cite = '<p><strong>If you use our service, please cite:</strong><br/><em>The Metagenomics RAST server - A public resource for the automatic phylogenetic and functional analysis of metagenomes</em> F. Meyer, D. Paarmann, M. D\'Souza, R. Olson , E. M. Glass, M. Kubal, T. Paczian , A. Rodriguez , R. Stevens, A. Wilke, J. Wilkening, R. A. Edwards<br/><em>BMC Bioinformatics 2008, 9:386
, [<a href="http://www.biomedcentral.com/1471-2105/9/386" target="_blank">article</a>]</em></p>';

  $tv->add_tab('Overview', '<div style="padding-left: 5px; padding-right: 5px; text-align: justify;"><p>The metagenome overview page provides basic information and a summary regarding the selected metagenome. Information includes project name, project description, metagenome name and unique id as well as sequence length and percent GC statistics. Histograms of sequence length and GC content is also provided. In order to provide a brief overview of the taxonomic distribution, a table is provided with domain distribution for RNA and protein based analysis.<p><p>The Overview is accessible through the menu via<br>
» Metagenome » <a href="?page=MetagenomeOverview&metagenome='.$metagenome.'">Overview</a></p></div>');

  $tv->add_tab('Phylogenetic Classification', '<div style="padding-left: 5px; padding-right: 5px; text-align: justify;"><strong>Phylogenetic Classification:</strong> We provide the taxonomic classification based on a number of different ribosomal RNA databases: <a href="http://greengenes.lbl.gov">GreenGenes</a>, <a href="http://rdp.cme.msu.edu/">RDP-II</a> and <a href="http://www.psb.ugent.be/rRNA">European ribosomal RNA database</a>. In addition we compute a taxonomic profile from the protein similarities found in our underlying SEED database. You can modify the parameters of the calculated profile including e-value, p-value , percent identity and minimum alignment length. This will allow you to refine the analysis to suit the sample and sequence characteristics of your metagenome. The SEED database provides an alternative way to identify taxonomies in the sample. Protein encoding genes are BLASTed against the SEED database and the taxonomy of the best hit is used to compile taxonomies of the sample.</p><p>Views are interactive and are in graphical (pie charts) and tabular form.</p><p>Available through the menu via<br/>&raquo; Metagenome &raquo; <a href="?page=MetagenomeProfile&dataset='.$self->data('dataset_select_metabolic')->[0].'&metagenome='.$metagenome.'">Sequence Profile</a> (then select Phylogenetic Profile)</p></div>');

  $tv->add_tab('Metabolic Analysis', '<div style="padding-left: 5px; padding-right: 5px; text-align: justify;"><p><strong>Metabolic Reconstruction:</strong> MG-RAST provides a functional classification for this metagenome using SEED Subsystem Technology. It gives you detailed view which metabolic functions or activities may be present in your sample. Views are interactive and are in graphical (pie charts) and tabular form. You can modify the parameters of the calculated profile including e-value, p-value , percent identity and minimum alignment length. This will allow you to refine the analysis to suit the sequence characteristics of your sample.</p><p>KEGG maps are also available using the metabolic functions computed by MG-RAST. Comparative tools are also available.</p><p>Available through the menu via<br/>&raquo; Metagenome &raquo; <a href="?page=MetagenomeProfile&dataset='.$self->data('dataset_select_metabolic')->[0].'&metagenome='.$metagenome.'">Sequence Profile</a> (then select Metabolic Profile)<br>&raquo; Compare Metagenomes &raquo; <a href="?page=Kegg&organism='.$metagenome.'">KEGG Map</a></p></div>');
  
  $tv->add_tab('Compare', '<div style="padding-left: 5px; padding-right: 5px; text-align: justify;"><strong>Compare Metagenomes:</strong>The comparative analysis tools can be used to align multiple metagenomes along the different classifications, e.g. show relationships of certain subsystems in different metagenomic samples. Three tools are available to perform comparisons:</p><p><strong>Heatmaps</strong> allow for metabolic or phylogentic comparison of two or more metagenomes. Absolute or normalized values can be used as well as setting the number of groups allowed. You can also modify the parameters of the calculated profile including e-value, p-value , percent identity and minimum alignment length.</p><p><strong>Recruitment Plot</strong> compares the metabolism of the metagenome to a individual bacterial organism. Again, parameters can be modified, like e-value, p-value , percent identity and minimum alignment length. This an alpha release and the tool provides qualitative results in graphical form with option to zoom in.</p><p><strong>KEGG</strong> maps and tables are provided as an alternative view of the metabolic reconstruction of the metagenome. It allows you to compare with multiple metagenomes at different levels within the metabolic pathway hierarchy. Absolute counts and percentages are provided.</p><p>Available through the menu via<br/>&raquo; Compare Metagenomes &raquo; <a href="?page=MetagenomeComparison&dataset='.$self->data('dataset_select_metabolic')->[0].'&metagenome='.$metagenome.'">Heatmap Comparison</a><br>&raquo; Compare Metagenomes &raquo; <a href="?page=MetagenomeRecruitmentPlot&metagenome='.$metagenome.'">Recruitment Plot</a><br>&raquo; Compare Metagenomes &raquo; <a href="?page=Kegg&organism='.$metagenome.'">KEGG Map</a></p></div>');

  return $intro . "<div id='description' style='display: none;'>" . $description . "</div>" . $tv->output . $cite;
}
